### ISSUE TITLE
 Login - User logged in with account A via magic link for account B if A logged in Web

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#23409]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Precondition: User is logged in with account A on Web
1. Open desktop app
2. Request magic link to sign in with account B
3. Copy magic link from email
4. Paste the link into the same browser with Web login as account A from precondition
5. Modify link to staging
6. Open the link
7. Click on "Allow" in prompt to open the link in desktop app

## Expected Result:
User is logged in with account B on Desktop after opening magic link for account B from email

## Actual Result:
User is logged in with account A with which they were logged in on Web after opening magic link for account B.
On production build same issue happens if user is logged in with account A on Web and they also open the email with magic link and click on the link in same the browser (without additionally modifying it to staging)

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS /  Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.44.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/b43371e0-6a30-4f68-bd93-217eb19ce7ab

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by s77rt** [comment]
Not overdue. This just went `External`.

**Comment by s77rt** [comment]
Not overdue. Still looking for proposals

**Comment by JKobrynski** [comment]
I found a potential solution. Proposal will be published soon.

**Comment by s77rt** [comment]
Still working on item, Melvin

**Comment by JKobrynski** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

We want to prevent users from being logged into the wrong account via magic link. Currently, if a user is trying to log into account B in the desktop app via magic link, while being logged into account A in the browser, the magic link will work, but it will log him into account A (the one from the browser) instead of B (from the desktop app).

### What is the root cause of that problem?

When the user is redirected from the web app to the desktop app via magic link, **shortLivedAuthToken** is attached to that url, and it's the root cause of this issue because it's a token for account A, not B.

The very first thing that happens after user is redirected to the desktop app, is authorization with **shortLivedAuthToken** in `<LogInWithShortLivedAuthTokenPage />` component just like it's done for every other deep link. After that, the app redirects the user to `<ValidateLoginPage />` (**exitTo** property) that is supposed to use the arguments from the magic link for authorization, but there is a condition that prevents that from happening:

```js
if (lodashGet(props, 'session.authToken')) {
    // If already signed in, do not show the validate code if not on web,
    // because we don't want to block the user with the interstitial page.
    Navigation.goBack(false);
}
```
And because of this condition, the magic link is omitted.


### What changes do you think we should make in order to solve the problem?

We could prevent the web app from attaching **shortLivedAuthToken** and **email** to the link that is supposed to be opened in the desktop app if it's a magic link. So in *Browser/index.web.js* we could modify the existing condition:

```js
if (email && shortLivedAuthToken) {
    params.set('email', email);
    params.set('shortLivedAuthToken', shortLivedAuthToken);
}
```

to look like this

```js
const isMagicLink = /\/v\/[0-9]{8}\/[0-9]{6}/.test(window.location.pathname)

if (email && shortLivedAuthToken && !isMagicLink) {
    params.set('email', email);
    params.set('shortLivedAuthToken', shortLivedAuthToken);
}
```

And then if it's a magic link, the desktop app won't receive any token at all, just the magic link.

### What alternative solutions did you explore? (Optional)

We could also add a condition in `<LogInWithShortLivedAuthTokenPage />` to check if the deep link that is handled is a magic link and if so, skip the step that authorizes the user with **shortLivedAuthToken** and proceed to redirection, it could look something like:

```js
const isMagicLink = /\/v\/[0-9]{8}\/[0-9]{6}/.test(exitTo)

// Modify existing condition
if (shortLivedAuthToken && !isMagicLink) {
    Session.signInWithShortLivedAuthToken(email, shortLivedAuthToken);
    return;
}
```

**Comment by s77rt** [comment]
@JKobrynski Thanks for the proposal. Your RCA is correct. The solution makes sense, but can we not pass the auth token in the first place if the url is a magic link? (instead of passing and ignoring it)

Also I'm not exactly sure if this is a bug so I asked on slack before moving forward https://expensify.slack.com/archives/C01GTK53T8Q/p1691434399687539

**Comment by JKobrynski** [comment]
@s77rt got it, I went through the Slack messages, and it seems like it's probably not a bug, but just in case here is an updated proposal:

## Proposal

### Please re-state the problem that we are trying to solve in this issue.

We want to prevent users from being logged into the wrong account via magic link. Currently, if a user is trying to log into account B in the desktop app via magic link, while being logged into account A in the browser, the magic link will work, but it will log him into account A (the one from the browser) instead of B (from the desktop app).

### What is the root cause of that problem?

When the user is redirected from the web app to the desktop app via magic link, **shortLivedAuthToken** is attached to that url, and it's the root cause of this issue because it's a token for account A, not B.

The very first thing that happens after user is redirected to the desktop app, is authorization with **shortLivedAuthToken** in `<LogInWithShortLivedAuthTokenPage />` component just like it's done for every other deep link. After that, the app redirects the user to `<ValidateLoginPage />` (**exitTo** property) that is supposed to use the arguments from the magic link for authorization, but there is a condition that prevents that from happening:

```js
if (lodashGet(props, 'session.authToken')) {
    // If already signed in, do not show the validate code if not on web,
    // because we don't want to block the user with the interstitial page.
    Navigation.goBack(false);
}
```
And because of this condition, the magic link is omitted.


### What changes do you think we should make in order to solve the problem?

We could prevent the web app from attaching **shortLivedAuthToken** and **email** to the link that is supposed to be opened in the desktop app if it's a magic link. So in *App.js* we could modify the existing condition:

```js
if (!currentUserAccountID) {
    Browser.openRouteInDesktopApp();
    return;
}
```

to look like this

```js
// Match any magic link (/v/<account id>/<6 digit code>)
const isMagicLink = /\/v\/\w+\/[0-9]{6}/.test(window.location.pathname)

// If the route that is being handled is a magic link, email and shortLivedAuthToken should not be attached to the url 
// to prevent signing into the wrong account
if (!currentUserAccountID || isMagicLink) {
    Browser.openRouteInDesktopApp();
    return;
}
```

And thanks to that, **shortLivedAuthToken** and **email** wouldn't be attached to the url and the app would just handle the magic link.

### What alternative solutions did you explore? (Optional)

We could prevent the web app from attaching **shortLivedAuthToken** and **email** to the link that is supposed to be opened in the desktop app if it's a magic link. So in *Browser/index.web.js* we could modify the existing condition:

```js
if (email && shortLivedAuthToken) {
    params.set('email', email);
    params.set('shortLivedAuthToken', shortLivedAuthToken);
}
```

to look like this

```js
const isMagicLink = /\/v\/[0-9]{8}\/[0-9]{6}/.test(window.location.pathname)

if (email && shortLivedAuthToken && !isMagicLink) {
    params.set('email', email);
    params.set('shortLivedAuthToken', shortLivedAuthToken);
}
```

And then if it's a magic link, the desktop app won't receive any token at all, just the magic link.

**Comment by JKobrynski** [comment]
@s77rt You would get logged in as the account that the magic link was generated for. So let's say you are already logged in as A on the web and try to log in as A on the desktop app via magic link, then you would be logged in as A. The difference being authorization through magic link rather than **shortLivedAuthToken**. I hope that answers your question!

**Comment by s77rt** [comment]
@JKobrynski Sorry that was not clear. Let me rephrase the question:
1. On Web you are logged in as account A
2. On Desktop you are logged in as account C
3. On Web you open a magic link for account B and continue as deep link in Desktop


In Desktop will you:
1. Stay logged in as account C
2. Get logged in as account B

**Comment by JKobrynski** [comment]
@s77rt I haven't tested this scenario yet, but I am pretty sure it would be account B. **shortLivedAuthToken** would not be attached, so there is no way to be signed in as A, magic link would refer to account B and therefore the desktop app would use it to sign in.

**Comment by s77rt** [comment]
@JKobrynski I'm sure we won't login to account A as the token won't be attached. I'm asking if we'd still logged in as account C because we are already logged in that account

**Comment by JKobrynski** [comment]
@s77rt with the current implementation, I think no. I will test it and let you know tomorrow

**Comment by JKobrynski** [comment]
@s77rt I just tested this scenario on my branch and got logged into account B

**Comment by s77rt** [comment]
@JKobrynski Thank you! I think that looks good to me. Discussing in Slack for one last confirmation before moving forward

**Comment by JKobrynski** [comment]
@s77rt the [PR](https://github.com/Expensify/App/pull/24207) is ready for review

**Comment by s77rt** [comment]
Looks like automation didn't work here. PR is merged and got deployed to production 2 weeks ago.

**Comment by MitchExpensify** [comment]
Oh great catch @s77rt. Let's manually apply the BZ steps and get payment out as a next step.

**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@s77rt] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@s77rt] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@s77rt] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@s77rt] Determine if we should create a regression test for this bug.
- [ ] [@s77rt] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@MitchExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MitchExpensify** [comment]
**Payment summary:**

Reporting: N/A
C: N/A (CallStack)
C+: $1000 @s77rt 

Look OK to the you @s77rt ?

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: https://github.com/Expensify/App/pull/16869
- The offending PR has been commented on: https://github.com/Expensify/App/pull/16869#discussion_r1316033897
- A discussion in #expensify-bugs has been started: I think this is one kind of a bug, no discussion is needed for this one
- Determine if we should create a regression test for this bug: No. I think [C2616813](https://expensify.testrail.io/index.php?/cases/view/2616813) is enough

**Comment by MitchExpensify** [comment]
Paid and contract ended - Thanks @s77rt !
